docs/website/docs/v1.x/dsl/middleware.md

```diff
@@ -3,4 +3,422 @@ sidebar_position: "8"
+
+This can lead to a lot of boilerplate clogging our neatly written endpoints affecting readability, thereby leading to maintenance costs.
+
+## Need for middlewares and handling "aspects"
+
+If we refer to Wikipedia for the definition of an "[Aspect](https://en.wikipedia.org/wiki/Aspect_(computer_programming))" we can glean the following points.
+
+* An aspect of a program is a feature linked to many other parts of the program (**_most common example logging_**)., 
+* But it is not related to the program's primary function (**_core business logic_**) 
+* An aspect crosscuts the program's core concerns (**_for example logging code intertwined with core business logic_**),  
+* Therefore, it can violate the principle of "separation of concerns" which tries to encapsulate unrelated functions. (**_Code duplication and maintenance nightmare_**)
+
+Or in short, aspect is a common concern required throughout the application, and its implementation could lead to repeated boilerplate code and in violation of the principle of separation of concerns.
+There is a paradigm in the programming world called [aspect-oriented programming](https://en.wikipedia.org/wiki/Aspect-oriented_programming) that aims for modular handling of these common concerns in an application. 
+
+Some examples of common "aspects" required throughout the application
+- logging,
+- timeouts (preventing long-running code)
+- retries (or handling flakiness for example while accessing third party APIs)
+- authenticating a user before using the REST resource (basic, or custom ones like OAuth / single sign-on, etc).
+
+This is where middleware comes to the rescue. 
+Using middlewares we can compose out-of-the-box middlewares (or our custom middlewares) to address the above-mentioned concerns using ++ and @@ operators as shown below.
+
+#### Cleaned up code using middleware to address cross-cutting concerns like auth, request/response logging, etc.
+```scala
+// compose basic auth, request/response logging, timeouts middlewares
+val composedMiddlewares = Middleware.basicAuth("user","pw") ++ 
+        Middleware.debug ++ 
+        Middleware.timeout(5 seconds) 
+
+private val app = Http.collectZIO[Request] {
+  case Method.GET -> !! / "users" / id =>
+    // core business logic  
+    dbService.lookupUsersById(id).map(Response.json(_.json))
+  case Method.GET -> !! / "users"    =>
+    // core business logic  
+    dbService.paginatedUsers(pageNum).map(Response.json(_.json))
+} @@ composedMiddlewares // attach composedMiddlewares to the app using @@
+```
+Observe how we gained following benefits by using middlewares
+* **Readability**: de-cluttering business logic.
+* **Modularity**: we can manage aspects independently without making changes in 100 places. For example, 
+  * replacing the logging mechanism from logback to log4j2 will require a change in one place, the logging middleware).
```

Comment:
Good point. Can talk about the Log4J usecase before when it was all boilerplaty.